ICT4DNews/app/src/main/java/at/ict4d/ict4dnews/server/Server.kt

```diff
@@ -23,23 +28,162 @@ import javax.inject.Inject
+                    .onErrorReturn { emptyList() }
+                )
+            } else if (blog.feedType == FeedType.RSS || blog.feedType == FeedType.WORDPRESS_COM) {
+                /*
+                requests.add(apiRSSService.getRssICT4DatNews(blog.url)
+                    .onErrorReturn { RSSFeed(null) }
+                )*/
+            }
+        }
+
+        return Single.zip(requests) { serverResult ->
+            Timber.d("Result: ${serverResult.size}")
+            Timber.d("Result: ${serverResult[0]}")
+
+            for (item in serverResult) {
+
+                val serverResultBlogList = item as ArrayList<*>
+
+                if (serverResultBlogList.isNotEmpty()) {
+                    val firstItem = serverResultBlogList.first()
+                    if (firstItem is SelfHostedWPPost) {
+                        handleSelfHostedWPBlogList(blogs, serverResultBlogList)
+                    } else if (firstItem is RSSFeed) {
+                        // TODO: set up rss feed response and save to DB
+                    }
+                }
+            }
+            rxEventBus.post(NewsRefreshDoneMessage())
+        }
+            .subscribeOn(Schedulers.io())
+            .observeOn(Schedulers.io())
+            .subscribe({
+                Timber.d("**** done: $it")
+            }, {
+                Timber.e("**** Error in downloading news from all active posts")
+                handleError(it, NewsRefreshDoneMessage())
+            })
+    }
+
+    private fun handleSelfHostedWPBlogList(databaseBlogList: List<Blog>, serverResultBlog: ArrayList<*>) {
+
+        if (serverResultBlog.isNotEmpty() && serverResultBlog.first() is SelfHostedWPPost) {
+
+            // set blog URL
+            val dbBlog = databaseBlogList.find { dbBlog -> (serverResultBlog.first() as SelfHostedWPPost).link.contains(dbBlog.url) }
+            serverResultBlog.map { b -> (b as SelfHostedWPPost).blogLink = dbBlog?.url ?: "" }
+
+            // Query for the authors
+            val serverAuthors: MutableList<WordpressAuthor> = mutableListOf()
+            for (post in serverResultBlog.distinctBy { (it as SelfHostedWPPost).serverAuthorID }) {
+                try {
+                    val author = apiJsonSelfHostedWPService.getJsonNewsAuthorByID(dbBlog?.url + "wp-json/wp/v2/users/${(post as SelfHostedWPPost).serverAuthorID}/").execute().body()
+
+                    author?.let {
+                        serverAuthors.add(author)
+                    }
+                } catch (e: Throwable) {
+                    Timber.e("Error in downloading an author from a self-hosted Wordpress blog", e)
+                    rxEventBus.post(ServerErrorMessage(R.string.http_exception_error_message, e))
+                    return
+                }
+            }
+
+            // Query for all media elements per each post
+            val serverMedia: MutableList<WordpressMedia> = mutableListOf()
+            for (post in serverResultBlog) {
```

Comment:
may be we use Kotlin Collections method `forEach` instead of traditional forEach like this `post.forEach{post-> post.something}`

Reply:
True that! I fixed it and added more stuff. Looks already quite good. RSS is still missing